src/gamepad/mapping.vala

```diff
@@ -0,0 +1,196 @@
+// This file is part of GNOME Games. License: GPLv3
+
+private class LibGamepad.Dpad : Object {
+	public InputType types[4];
+	public int values[4];
+	public int axisval[2];
+
+	public Dpad () {
+		axisval[0] = axisval[1] = 0;
+		types[0] = types[1] = types[2] = types[3] = InputType.INVALID;
+	}
+}
+
+public class LibGamepad.Mapping : Object {
+	private InputType[] buttons_type;
+	private int[] buttons_value;
+	private InputType[] axes_type;
+	private int[] axes_value;
+	private Dpad[] dpads;
+
+	public Mapping.from_sdl_string (string? mapping_string) throws MappingError {
+		if (mapping_string == null || mapping_string == "")
+			throw new MappingError.NOT_A_MAPPING ("The mapping string can't be null nor empty.");
+
+		var mappings = mapping_string.split (",");
+		foreach (var mapping in mappings) {
+			if (mapping.split (":").length == 2) {
+				var str = mapping.split (":")[0];
+				var real = mapping.split (":")[1];
+				var type = map_type (str);
+				if (type == InputType.INVALID) {
+					if (str != "platform") debug ("Invalid token : %s", str);
+					continue;
```

Comment:
Newline before `continue`, `break`, `throw` and `return` everywhere is they are not the first (and hence only) statement of the block.